test/e2e/reboot.go

```diff
@@ -95,28 +114,44 @@ var _ = Describe("Reboot", func() {
 })
 
 func testReboot(c *client.Client, rebootCmd string) {
+	// These tests requires SSH to nodes, so the provider check should be identical to there
+	// (the limiting factor is the implementation of util.go's getSigner(...)).
+
+	// Cluster must support node reboot
+	SkipUnlessProviderIs(providersWithSSH...)
+
 	// Get all nodes, and kick off the test on each.
 	nodelist, err := listNodes(c, labels.Everything(), fields.Everything())
 	if err != nil {
 		Failf("Error getting nodes: %v", err)
 	}
-	result := make(chan bool, len(nodelist.Items))
-	for _, n := range nodelist.Items {
-		go rebootNode(c, testContext.Provider, n.ObjectMeta.Name, rebootCmd, result)
-	}
+	result := make([]bool, len(nodelist.Items))
+	wg := sync.WaitGroup{}
+	wg.Add(len(nodelist.Items))
 
-	// Wait for all to finish and check the final result.
 	failed := false
-	// TODO(a-robinson): Change to `for range` syntax and remove logging once
-	// we support only Go >= 1.4.
-	for _, n := range nodelist.Items {
-		if !<-result {
-			Failf("Node %s failed reboot test.", n.ObjectMeta.Name)
-			failed = true
-		}
+	for ix := range nodelist.Items {
+		go func(ix int) {
+			defer wg.Done()
+			n := nodelist.Items[ix]
+			result[ix] = rebootNode(c, testContext.Provider, n.ObjectMeta.Name, rebootCmd)
+			if !result[ix] {
+				failed = true
+			}
+		}(ix)
 	}
+
+	// Wait for all to finish and check the final result.
+	wg.Wait()
+
 	if failed {
-		Failf("Test failed; at least one node failed to reboot in the time given.")
+		for ix := range nodelist.Items {
+			n := nodelist.Items[ix]
+			if !result[ix] {
+				Failf("Node %s failed reboot test.", n.ObjectMeta.Name)
```

Comment:
this will abort at the first failure; if we want to see all failing nodes, this probably needs to be Logf instead.